rest-api/dao/biobank_order_dao.py

```diff
@@ -339,4 +391,140 @@ def to_client_json(self, model):
+    self._update_identifier_history(session, order)
+    self._update_sample_history(session, order)
+
+  def update_with_patch(self, id_, resource, expected_version):
+    """creates an atomic patch request on an object. It will fail if the object
+    doesn't exist already, or if obj.version does not match the version of the existing object.
+    May modify the passed in object."""
+    with self.session() as session:
+      obj = self.get_with_children_in_session(session, id_, for_update=True)
+      return self.patch_update_with_session(session, obj, resource, expected_version)
+
+  def _do_update_with_patch(self, session, order, resource):
+    order.lastModified = clock.CLOCK.now()
+    order.logPosition = LogPosition()
+    order.version += 1
+    if resource['status'].lower() == 'cancelled':
+      order.amendedReason = resource['amendedReason']
+      order.cancelledUsername = resource['cancelledInfo']['author']['value']
+      order.cancelledSiteId = get_site(resource['cancelledInfo'])
+      order.cancelledTime = clock.CLOCK.now()
+      order.orderStatus = BiobankOrderStatus.CANCELLED
+    elif resource['status'].lower() == 'restored':
+      order.amendedReason = resource['amendedReason']
+      order.restoredUsername = resource['restoredInfo']['author']['value']
+      order.restoredSiteId = get_site(resource['restoredInfo'])
+      order.restoredTime = clock.CLOCK.now()
+      order.orderStatus = BiobankOrderStatus.UNSET
+    else:
+      raise BadRequest('status must be restored or cancelled for patch request.')
+
+    self._update_history(session, order)
+    self._update_identifier_history(session, order)
+    self._update_sample_history(session, order)
+    super(BiobankOrderDao, self)._do_update(session, order, resource)
+    return order
+
+  def _validate_patch_update(self, session, model, resource, expected_version):
+    required_cancelled_fields = ['amendedReason', 'cancelledInfo', 'status']
+    required_restored_fields = ['amendedReason', 'restoredInfo', 'status']
+    if 'status' not in resource:
+      raise BadRequest('status of cancelled/restored is required')
+
+    if resource['status'] == 'cancelled':
+      for field in required_cancelled_fields:
+        if field not in resource:
+          raise BadRequest('%s is required for a cancelled biobank order' % field)
+      if 'site' not in resource['cancelledInfo'] or 'author' not in resource['cancelledInfo']:
+        raise BadRequest('author and site are required for cancelledInfo')
+
+    elif resource['status'] == 'restored':
+      for field in required_restored_fields:
+        if field not in resource:
+          raise BadRequest('%s is required for a cancelled biobank order' % field)
+      if 'site' not in resource['restoredInfo'] or 'author' not in resource['restoredInfo']:
+        raise BadRequest('author and site are required for restoredInfo')
+
+    super(BiobankOrderDao, self)._validate_patch_update(session, model, resource, expected_version)
```

Comment:
See above; we can just inline the logic from base_dao.validate_patch_update here.

We should also validate that you can only cancel an order that isn't canceled, and only restore an order that is.